libs/mbedtls/mbedtls_stubs.c

```diff
@@ -0,0 +1,307 @@
+	r = mbedtls_entropy_func(PVOID(data), String_val(output), Int_val(len));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_entropy_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_entropy_context* entropy = malloc(sizeof(mbedtls_entropy_context));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)entropy);
+	mbedtls_entropy_init(entropy);
+	CAMLreturn(obj);
+}
+
+// Config
+
+CAMLprim void ml_mbedtls_ssl_conf_authmode(value conf, value authmode) {
+	CAMLparam2(conf, authmode);
+	mbedtls_ssl_conf_authmode(CFG(conf), Int_val(authmode));
+	CAMLreturn0;
+}
+
+CAMLprim void ml_mbedtls_ssl_conf_ca_chain(value conf, value ca_chain, value ca_crl) {
+	CAMLparam3(conf, ca_chain, ca_crl);
+	mbedtls_ssl_conf_ca_chain(CFG(conf), CERT(ca_chain), NULL /* TODO */);
+	CAMLreturn0;
+}
+
+CAMLprim value ml_mbedtls_ssl_config_defaults(value conf, value endpoint, value transport, value preset) {
+	CAMLparam4(conf, endpoint, transport, preset);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_config_defaults(CFG(conf), Int_val(endpoint), Int_val(transport), Int_val(preset));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim void ml_mbedtls_ssl_config_free(value conf) {
+	CAMLparam1(conf);
+	mbedtls_ssl_config_free(CFG(conf));
+	CAMLreturn0;
+}
+
+CAMLprim value ml_mbedtls_ssl_config_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_ssl_config* conf = malloc(sizeof(mbedtls_ssl_config));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)conf);
+	mbedtls_ssl_config_init(conf);
+	CAMLreturn(obj);
+}
+
+CAMLprim void ml_mbedtls_ssl_conf_rng(value conf, value p_rng) {
+	CAMLparam2(conf, p_rng);
+	mbedtls_ssl_conf_rng(CFG(conf), mbedtls_ctr_drbg_random, PVOID(p_rng));
+	CAMLreturn0;
+}
+
+// Ssl
+
+CAMLprim void ml_mbedtls_ssl_free(value ssl) {
+	CAMLparam1(ssl);
+	mbedtls_ssl_free(SSL(ssl));
+	CAMLreturn0;
+}
+
+CAMLprim value ml_mbedtls_ssl_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_ssl_context* ssl = malloc(sizeof(mbedtls_ssl_context));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)ssl);
+	mbedtls_ssl_init(ssl);
+	CAMLreturn(obj);
+}
+
+CAMLprim value ml_mbedtls_ssl_handshake(value ssl) {
+	CAMLparam1(ssl);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_handshake(SSL(ssl));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_ssl_read(value ssl, value buf, value pos, value len) {
+	CAMLparam4(ssl, buf, pos, len);
+	CAMLlocal1(r);
+	r = mbedtls_ssl_read(SSL(ssl), String_val(buf) + Int_val(pos), Int_val(len));
+	CAMLreturn(Val_int(r));
+}
+
+int bio_write_cb(void* ctx, const unsigned char* buf, size_t len) {
+	CAMLparam0();
+	CAMLlocal3(r, s, vctx);
+	vctx = (value)ctx;
+	s = caml_alloc_string(len);
+	memcpy(String_val(s), buf, len);
+	r = caml_callback2(Field(vctx, 1), Field(vctx, 0), s);
+	return Int_val(r);
+}
+
+int bio_read_cb(void* ctx, unsigned char* buf, size_t len) {
+	CAMLparam0();
+	CAMLlocal3(r, s, vctx);
+	vctx = (value)ctx;
+	s = caml_alloc_string(len);
+	r = caml_callback2(Field(vctx, 2), Field(vctx, 0), s);
```

Comment:
This is actually about bytes, but I had issues with 4.03 OCaml not having `Bytes_val` or something. It's probably possible to clean this up though.

Reply:
Actually, this is fine because both `string` and `bytes` have the same native representation. We could instead use `caml_alloc(len, String_tag)`, but there's no reason to do that.